esp/logging/logginglib/modularlogagent.cpp

```diff
@@ -0,0 +1,1526 @@
+                        format += 1; // consumed one additional character
+                    }
+                    else
+                        result = false;
+                    break;
+                }
+            }
+            format++;
+        }
+    }
+    else
+        result = false;
+    return result;
+}
+inline bool checkTimestampFormat(const char* format)
+{
+    return checkTimestampFormat(format, nullptr);
+}
+
+/**
+ * Create a timestamp string based on a given time. The function uses strftime with a non-literal
+ * format string. This normally generates a warning, "-Wformat-nonliteral", that is treated as an
+ * error. This warning must be temporarily ignored for the function to compile.
+ */
+inline StringBuffer& makeTimestamp(struct tm& when, const char* format, StringBuffer& timestamp)
+{
+    size_t capacity = 100;
+    char* buffer = (char*)malloc(capacity);
+    size_t length = 0;
+    byte tries = 0;
+    while (buffer && (length = strftime(buffer, capacity, format, &when)) == 0 && ++tries <= 3)
+    {
+        capacity *= 2;
+        char* tmp = (char*)realloc(buffer, capacity);
+        if (!tmp)
+        {
+            free(buffer);
+            buffer = nullptr;
+        }
+        buffer = tmp;
+    }
+    if (length)
+        timestamp.append(buffer);
+    free(buffer);
+    return timestamp;
+}
+
+/**
+ * Create a timestamp string based on the current UTC time.
+ */
+inline StringBuffer& makeTimestamp(const char* format, StringBuffer& timestamp)
+{
+    time_t nowTime;
+    time(&nowTime);
+    struct tm nowTm;
+    struct tm* nowInfo = gmtime_r(&nowTime, &nowTm);
+    return makeTimestamp(*gmtime_r(&nowTime, &nowTm), format, timestamp);
+}
+
+#if defined(__clang__) || defined(__GNUC__)
+#pragma GCC diagnostic pop
+#endif
+
+//////////////////// CModule ////////////////////
+
+LogMsgDetail CModule::tracePriorityLimit(const LogMsgCategory& category) const
+{
+    return (m_inheritTracePriorityLimit ? (getEspLogLevel() * 10) : m_tracePriorityLimit);
+}
+
+const char* CModule::traceId() const
+{
+    return m_traceId;
+}
+
+void CModule::traceOutput(const LogMsgCategory& category, const char* format, va_list& arguments) const
+{
+    StringBuffer baseMsg;
+    baseMsg.valist_appendf(format, arguments);
+    LOG(category, "%s: %s", traceId(), baseMsg.str());
+}
+
+bool CModule::configure(const IPTree& configuration, const CModuleFactory& factory)
+{
+    static const int INHERIT_TRACE_PRIORITY_LIMIT = -1;
+    bool result = true;
+
+    // Neither CModule nor TModule<> are logging components, but implementations of IModule will be.
+    // Enable the common base(s) to use the derived logging methods.
+    m_self = dynamic_cast<ITraceLoggingComponent*>(this);
+
+    m_traceId.append(configuration.queryName());
+    const char* name = configuration.queryProp(propName);
+    const char* module = configuration.queryProp(IModule::propFactoryKey);
+    if (!isEmptyString(name))
+    {
+        if (module)
```

Comment:
Yes. If omitted from the configuration, it will be omitted from the ID. If present in the configuration, even if empty, it will be included in the ID. I've chosen to have the ID accurately reflect the configured content, even for an empty value.